### PR TITLE
NXP-25012 s3 bucket prefix + refactoring

### PR DESCRIPTION
**Pull request by andreinechaev**

### CODE REVIEW

File: nuxeo-lambda-image-conversion/src/main/java/org/nuxeo/lambda/image/conversion/listener/LambdaResponseListener.java [hunk 1]
```diff
@@ -38,8 +30,19 @@
 import org.nuxeo.ecm.core.work.api.WorkManager;
 import org.nuxeo.ecm.platform.picture.PictureViewsGenerationWork;
 import org.nuxeo.lambda.core.LambdaService;
+import static org.nuxeo.lambda.core.LambdaService.LAMBDA_RESPONSE_KEY;
+import static org.nuxeo.lambda.core.LambdaService.PARAMETERS_KEY;
+import static org.nuxeo.lambda.image.conversion.common.Constants.DOC_ID_PROP;
+import static org.nuxeo.lambda.image.conversion.common.Constants.REPOSITORY_PROP;
```
Comment (kevinleturc) [inline nuxeo-lambda-image-conversion/src/main/java/org/nuxeo/lambda/image/conversion/listener/LambdaResponseListener.java]: Imports are not well organised. We should have:
- static
- java
- javax
- org

File: nuxeo-lambda-image-conversion/src/main/java/org/nuxeo/lambda/image/conversion/listener/PictureCreatedListener.java [hunk 2]
```diff
@@ -17,59 +17,54 @@
...
+import static org.nuxeo.lambda.image.conversion.common.Constants.BUCKET_PROP;
+import static org.nuxeo.lambda.image.conversion.common.Constants.CONVERSIONS_SIZES;
+import static org.nuxeo.lambda.image.conversion.common.Constants.DEFAULT_LAMBDA_PICTURE_NAME;
+import static org.nuxeo.lambda.image.conversion.common.Constants.DIGEST_PROP;
+import static org.nuxeo.lambda.image.conversion.common.Constants.DOC_ID_PROP;
+import static org.nuxeo.lambda.image.conversion.common.Constants.FILENAME;
+import static org.nuxeo.lambda.image.conversion.common.Constants.LIFECYCLE;
+import static org.nuxeo.lambda.image.conversion.common.Constants.MIME_TYPE;
+import static org.nuxeo.lambda.image.conversion.common.Constants.NUXEO_LAMBDA_IMAGE_PROP;
+import static org.nuxeo.lambda.image.conversion.common.Constants.REPOSITORY_PROP;
```
Comment (kevinleturc) [inline nuxeo-lambda-image-conversion/src/main/java/org/nuxeo/lambda/image/conversion/listener/PictureCreatedListener.java:36]: Import order.

File: nuxeo-lambda-image-conversion/src/test/java/org/nuxeo/lambda/image/conversion/TestPictureViewCreateWork.java [hunk 3]
```diff
@@ -42,19 +31,26 @@
...
 import org.nuxeo.runtime.transaction.TransactionHelper;
 
+import javax.inject.Inject;
+import java.io.File;
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
```
Comment (kevinleturc) [inline nuxeo-lambda-image-conversion/src/test/java/org/nuxeo/lambda/image/conversion/TestPictureViewCreateWork.java]: Import order.

File: nuxeo-lambda-image-conversion/src/test/java/org/nuxeo/lambda/image/conversion/common/LambdaFeature.java [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+import org.nuxeo.runtime.api.Framework;
+import org.nuxeo.runtime.test.runner.Deploy;
+import org.nuxeo.runtime.test.runner.Features;
+import org.nuxeo.runtime.test.runner.FeaturesRunner;
+import org.nuxeo.runtime.test.runner.SimpleFeature;
+
+import java.util.Properties;
+
+@Features(PlatformFeature.class)
+@Deploy({ "org.nuxeo.lambda.core", "org.nuxeo.lambda.image.conversion", "org.nuxeo.ecm.platform.picture.api",
```
Comment (kevinleturc) [inline nuxeo-lambda-image-conversion/src/test/java/org/nuxeo/lambda/image/conversion/common/LambdaFeature.java]: Following [NXP-24366](https://jira.nuxeo.com/browse/NXP-24366), we can now express several `@Deploy` on the same entity. We have changed guidelines and now use several annotation instead of array.

**Review by kevinleturc — APPROVED** [review]
Just some format, 👍

File: nuxeo-lambda-image-conversion/src/main/java/org/nuxeo/lambda/image/conversion/listener/LambdaResponseListener.java [hunk 5]
```diff
@@ -58,10 +61,18 @@
     public void handleEvent(Event event) {
         EventContext ctx = event.getContext();
         Map<String, Serializable> params = (Map<String, Serializable>) ctx.getProperty(PARAMETERS_KEY);
+        if (params == null || params.isEmpty()) {
+            log.warn("Couldn't get parameters from the context");
```
Comment (efge) [inline nuxeo-lambda-image-conversion/src/main/java/org/nuxeo/lambda/image/conversion/listener/LambdaResponseListener.java]: This (and others) is probably not useful as a WARN. It think it should be DEBUG only.

File: nuxeo-lambda-image-conversion/src/main/java/org/nuxeo/lambda/image/conversion/listener/PictureCreatedListener.java [hunk 6]
```diff
@@ -90,7 +85,7 @@ public void handleEvent(Event event) {
             Blob blob = (Blob) content.getValue();
 
             if (blob == null) {
-                log.debug("Attempt to create Picture View with no data");
+                log.warn("Attempt to create Picture View with no data");
```
Comment (efge) [inline nuxeo-lambda-image-conversion/src/main/java/org/nuxeo/lambda/image/conversion/listener/PictureCreatedListener.java]: I'd keep it DEBUG.

File: nuxeo-lambda-image-conversion/src/test/java/org/nuxeo/lambda/image/conversion/TestLambdaPictureConversion.java [hunk 7]
```diff
@@ -77,7 +59,7 @@ public void clear() {
         MemLambdaCaller.repositoryName = session.getRepositoryName();
     }
 
-    @Test
+    @Test(timeout = 30 * 1000)
```
Comment (efge) [inline nuxeo-lambda-image-conversion/src/test/java/org/nuxeo/lambda/image/conversion/TestLambdaPictureConversion.java]: You can write `30_000` now since Java 8.

File: nuxeo-lambda-image-conversion/src/test/java/org/nuxeo/lambda/image/conversion/TestLambdaPictureConversion.java [hunk 8]
```diff
@@ -21,48 +21,30 @@
...
+import java.util.List;
+import java.util.Map;
+
 @RunWith(FeaturesRunner.class)
-@Features({ PlatformFeature.class })
-@Deploy({ "org.nuxeo.lambda.core", "org.nuxeo.lambda.image.conversion", "org.nuxeo.ecm.platform.picture.api",
-                "org.nuxeo.ecm.platform.picture.core", "org.nuxeo.ecm.platform.tag", "org.nuxeo.ecm.core.cache",
-                "org.nuxeo.ecm.automation.core" })
-@LocalDeploy("org.nuxeo.lambda.image.conversion.test:OSGI-INF/lambda-picture-config-override-contrib.xml")
+@Features({ LambdaFeature.class })
```
Comment (efge) [inline nuxeo-lambda-image-conversion/src/test/java/org/nuxeo/lambda/image/conversion/TestLambdaPictureConversion.java]: As there's only one feature you can write it without braces: `@Features(LambdaFeature.class)` (several places)

File: nuxeo-lambda-image-conversion/src/test/java/org/nuxeo/lambda/image/conversion/common/LambdaFeature.java [hunk 9]
```diff
@@ -0,0 +1,48 @@
+/*
+ * (C) Copyright 2006-2018 Nuxeo (http://nuxeo.com/) and others.
```
Comment (efge) [inline nuxeo-lambda-image-conversion/src/test/java/org/nuxeo/lambda/image/conversion/common/LambdaFeature.java]: If this is a new file use the year it was created, not 2006.

File: nuxeo-lambda-image-conversion/src/test/java/org/nuxeo/lambda/image/conversion/common/MemLambdaCaller.java [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+ * limitations under the License.
+ *
+ *
+ * Contributors:
+ *     anechaev
+ */
+package org.nuxeo.lambda.image.conversion.common;
+
+import org.codehaus.jettison.json.JSONArray;
+import org.codehaus.jettison.json.JSONObject;
```
Comment (efge) [inline nuxeo-lambda-image-conversion/src/test/java/org/nuxeo/lambda/image/conversion/common/MemLambdaCaller.java]: Hmm I see it was already used in other files of nuxeo-lambda previously, but please don't use `org.codehaus.jettison`. Our JSON library should be standardized on Jackson 2 only (could be refactored in a separate commit in all of nuxeo-lambda).

File: nuxeo-lambda-image-conversion/src/test/java/org/nuxeo/lambda/image/conversion/common/MemLambdaCaller.java [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+import org.nuxeo.ecm.core.blob.binary.Binary;
+import org.nuxeo.ecm.core.blob.binary.BinaryManager;
+import org.nuxeo.lambda.core.AbstractLambdaCaller;
+import org.nuxeo.lambda.core.LambdaService;
+import org.nuxeo.runtime.api.Framework;
+
+import java.util.Arrays;
+import java.util.concurrent.Executors;
+
+public class MemLambdaCaller extends AbstractLambdaCaller {
```
Comment (efge) [inline nuxeo-lambda-image-conversion/src/test/java/org/nuxeo/lambda/image/conversion/common/MemLambdaCaller.java]: Add a small Javadoc explaining what this test class does

File: nuxeo-lambda-image-conversion/src/test/java/org/nuxeo/lambda/image/conversion/common/MemLambdaCaller.java [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+
+    public static volatile String repositoryName = null;
+
+    public static volatile boolean completed = false;
+
+    public static volatile Exception exception = null;
```
Comment (efge) [inline nuxeo-lambda-image-conversion/src/test/java/org/nuxeo/lambda/image/conversion/common/MemLambdaCaller.java]: Remove `= null` and `= false` which are redundant.

**Review by efge — CHANGES_REQUESTED** [review]

File: nuxeo-lambda-image-conversion/src/test/java/org/nuxeo/lambda/image/conversion/TestLambdaPictureConversion.java [hunk 13]
```diff
@@ -36,17 +43,13 @@
...
-import java.util.List;
-import java.util.Map;
-
 @RunWith(FeaturesRunner.class)
-@Features({ LambdaFeature.class })
+@Features(LambdaFeature.class)
 @Deploy("org.nuxeo.lambda.image.conversion.test:OSGI-INF/lambda-picture-config-override-contrib.xml")
 public class TestLambdaPictureConversion {
 
+    public static final int _30_SEC = 30_000;
```
Comment (efge) [inline nuxeo-lambda-image-conversion/src/test/java/org/nuxeo/lambda/image/conversion/TestLambdaPictureConversion.java]: This is overkill but if you really want to use a separate constant for it it should be name something about its function and not about its value. Like `TEST_TIMEOUT`

**Review by efge — APPROVED** [review]

File: nuxeo-lambda-image-conversion/src/test/java/org/nuxeo/lambda/image/conversion/common/MemLambdaCaller.java (hunk 10 above)
Comment (andreinechaev) [reply nuxeo-lambda-image-conversion/src/test/java/org/nuxeo/lambda/image/conversion/common/MemLambdaCaller.java]: https://jira.nuxeo.com/browse/NXP-25023

File: nuxeo-lambda-image-conversion/src/test/java/org/nuxeo/lambda/image/conversion/TestLambdaPictureConversion.java (hunk 13 above)
Comment (andreinechaev) [reply nuxeo-lambda-image-conversion/src/test/java/org/nuxeo/lambda/image/conversion/TestLambdaPictureConversion.java]: My intention was to explicitly show the timeout. I believe that `30 * 1000` is clearer than `30_000` when we are talking about time. Anyway, it's changed to `30_000` now.
